pkg/resource/manager.go

```diff
@@ -125,12 +142,216 @@ func (m *manager) BatchCreateGenericResources(ctx context.Context, request *pb.B
+		DocumentIds: versionIds,
+		Index:       m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+	})
+	if err != nil {
+		return err
+	}
+
+	// versions that don't exist need to be created
+	// versions that do exist may need to be updated
+	var bulkItems []*esutil.BulkRequestItem
+	for i, versionId := range versionIds {
+		version := genericResourceVersions[versionId]
+		bulkItem := &esutil.BulkRequestItem{
+			Operation:  esutil.BULK_INDEX,
+			Message:    version,
+			DocumentId: versionId,
+			Join: &esutil.EsJoin{
+				Field:  genericResourceDocumentJoinField,
+				Name:   genericResourceVersionRelationName,
+				Parent: genericResourceIdFromVersionId(versionId),
+			},
+		}
+
+		if multiGetResponse.Docs[i].Found {
+			// if the version already exists, we may have to update(index) it.
+			// if we have a list of names, update the existing version with the new one
+			// otherwise, we have nothing to do here
+			if len(version.Names) != 0 {
+				bulkItem.Operation = esutil.BULK_INDEX
+				log.Debug("updating generic resource version", zap.Any("version", version))
+			} else {
+				bulkItem = nil
+			}
+		} else {
+			// if the version doesn't exist, we need to create it
+			// before creating the version, add a timestamp if it doesn't exist
+			if version.Created == nil {
+				version.Created = timestamppb.Now()
+			}
+
+			bulkItem.Operation = esutil.BULK_CREATE
+			log.Debug("creating generic resource version", zap.Any("version", version))
+		}
+
+		if bulkItem != nil {
+			bulkItems = append(bulkItems, bulkItem)
+		}
+	}
+
+	// perform bulk create / update
+	var bulkItemErrors []error
+	if len(bulkItems) != 0 {
+		bulkResponse, err := m.esClient.Bulk(ctx, &esutil.BulkRequest{
+			Index:   m.indexManager.AliasName(genericResourcesDocumentKind, ""),
+			Refresh: m.esConfig.Refresh.String(),
+			Items:   bulkItems,
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, item := range bulkResponse.Items {
+			var itemError error
+			if item.Create != nil && item.Create.Error != nil {
+				itemError = fmt.Errorf("error creating generic resource [%d] %s: %s", item.Create.Status, item.Create.Error.Type, item.Create.Error.Reason)
+			} else if item.Index != nil && item.Index.Error != nil {
+				itemError = fmt.Errorf("error re-indexing generic resource [%d] %s: %s", item.Create.Status, item.Create.Error.Type, item.Create.Error.Reason)
+			}
+			if itemError != nil {
+				bulkItemErrors = append(bulkItemErrors, itemError)
+			}
+		}
+	}
+
+	if len(bulkItemErrors) > 0 {
+		return fmt.Errorf("failed to create/update some generic resource versions: %v", bulkItemErrors)
+	}
+
+	return nil
+}
+
+// ListGenericResourceVersions handles the main logic for fetching versions associated with a generic resource.
+func (m *manager) ListGenericResourceVersions(ctx context.Context, request *pb.ListGenericResourceVersionsRequest) (*pb.ListGenericResourceVersionsResponse, error) {
+	return nil, nil
+}
```

Comment:
I'm going to leave the RPC for this unimplemented until a future PR.  for now, we're at least storing the versions correctly.  some additional changes have to be made to list them correctly.